add UtilsTests for Reflections and Math

bacchuongdaungo opened the pull request:
fix wrong-place utils test

bacchuongdaungo committed:
add UtilsTests for Reflections

lhmcgann committed:
Formatting edits

lhmcgann reviewed (changes requested):
Great start! Check code coverage. And if you could edit the formatting a bit (I started it) and add doc-comments (for test methods and methods tested) and in-line comments for anything more complicated, that would be great

bacchuongdaungo committed:
editted the formatting by adding doc-comments for test methods and methods tested and in-line comments for anything more complicated

lhmcgann commented on docs/images/association_arrows-techvidvan.png:
Are you able to undo all the changes to the png files? I think they were already made in the release branch. If not just lmk, it's fine, but if you can it will keep our git history cleaner

bacchuongdaungo replied:
The remaining uncovered code are from private classes for simulating test cases. Since they are only for testing and exist only in the test file, I am not sure if it's necessary to achieve full code coverage on those. I can still fix it if it's necessary.

lhmcgann commented on tests/EStimLibrary.UnitTests/Core/UtilsTests.cs:
```diff
@@ -1,123 +1,212 @@
+    /// Tests that GetAvailableTypes returns types assignable from the specified base type.
+    /// </summary>
     [Fact]
     public void GetAvailableTypes_ShouldReturnTypes_WhenTypeExists()
```
This is a style note, but for convention, the "ShouldXXX" component of test method names comes last. Convention is <MethodTested>_<Scenario>_Should<ExpectedOutput>()

lhmcgann commented on tests/EStimLibrary.UnitTests/Core/UtilsTests.cs:
```diff
@@ -130,6 +219,13 @@ public SampleClass(int param1 = 0)
+    /// <summary>
+    /// Factory interface for generic type testing.
+    /// </summary>
     public interface IFactory<T> { }
```
The library already has an IFactory<T> interface you can use instead of creating a new one here if you want, just to avoid confusion. Or, if you want to keep this test class referencing the fewest number of external classes (which is a valid design argument), just name this to something else to avoid confusion, e.g., MyGenericParameterizedClass<> (even though that's tedious) or something like that

lhmcgann replied:
especially since the IsManufacturableProduct(), TryFactoryCreate(), and similar functions refer to the EStimLibrary IFactory<> type

bacchuongdaungo replied:
I have changed the name of the private interfaces inside the test class.

lhmcgann reviewed (changes requested):
The existing tests are good, but a few main things:
1) check code coverage! There are still some fully untested functions, and many functions with only partial coverage. See the report for line-by-line coverage highlighting.
2) The doc comments in the test class are great! Just please add doc comments to all the Utils.cs source reflection functions as well!
3) You can delete the GetAvailableStimulatorTypes() utils function in Utils.cs, as well as the associated TODO comment, and don't worry about testing that or letting it mess up your coverage report

bacchuongdaungo committed:
Deleted GetAvailableStimulatorTypes(), added new test cases and added doc comments for them, refactored temp interfaces to avoid confusion, uniformed test name style

bacchuongdaungo committed:
Deleted GetAvailableStimulatorTypes(), added new test cases and added doc comments for them, refactored temp interfaces to avoid confusion, uniformed test name style

bacchuongdaungo committed:
Merge branch 'release-v2.0' into test/utils-2

bacchuongdaungo committed:
adjusted UtilsTests.cs to pass CI/CD pipeline

bacchuongdaungo committed:
Merge branch 'test/utils-2' of https://github.com/lhmcgann/EStimLibrary into test/utils-2

bacchuongdaungo committed:
Merge branch 'release-v2.0' into test/utils-2

bacchuongdaungo committed:
fix to pass ci/cd pipeline

lhmcgann reviewed (changes requested):
@bacchuongdaungo Glad the CI/CD fix was easy, and nice using the IFactory we already have. Check code coverage report for the following Reflection functions:

1. IsAssignableFromType()
2. IsGenericAssignableFrom()
3. GetAvailableGenericTypes()
4. IsManufacturableProduct()

And make sure there are doc comments `///` for all the reflection and math functions
(I'm including the first few functions in the Math region in this PR since you already included tests for them)

lhmcgann committed:
Removed old stimulator utils function

bacchuongdaungo committed:
added test cases for IsManufacturableProduct() and doc comments.

bacchuongdaungo committed:
Merge branch 'test/utils-2' of https://github.com/lhmcgann/EStimLibrary into test/utils-2

bacchuongdaungo committed:
Merge branch 'release-v2.0' into test/utils-2

lhmcgann committed:
Added tests; cleanup

- IsAssignableFrom test cases; some FAILing
- comments and other cleanup

bacchuongdaungo committed:
Refactor Utils.IsAssignableFromType and update tests

- Fixed generic assignability by scanning base classes and interfaces,
  and by checking type parameters.
- Updated dummy type hierarchies and test expectations in UtilsTests.cs.
- All tests now pass.

bacchuongdaungo committed:
- Added tests for interactive I/O and utility functions (e.g. TryParseEnumerableOfStrings, DictOfSetsToString, SelectFromList, ConstructWithUserInputParams, etc.)
- Refined reflection and assignability tests
- Minor cleanup and documentation updates

bacchuongdaungo committed:
Added few more test cases to push to 100%

bacchuongdaungo committed:
- Added new test cases for uncovered methods:
  • IsGenericAssignableFrom for non-generic test types
  • AreTypeParametersCompatible with covariant and incompatible types
  • ReadJSON error path when file is missing
  • EnumerableToString with empty enumerable
  • SelectFromList with invalid then valid input
  • SelectPort (if available)
  • RequestConstructorParameterValues using a multi-parameter dummy constructor

bacchuongdaungo committed:
Push IsGenericAssignableFrom test coverage to 100%

bacchuongdaungo committed:
- Added comprehensive tests for `IsAssignableFromType`, `IsGenericAssignableFrom`, and `RequestFactoryCreateParamValues`.
- Improved existing tests by removing dummy implementations and leveraging actual IDataLimits implementations.
- Fixed TypeConverter issues during testing sequences and collection parsing.
